Modernize parts of codebase and remove deprecated code-paths

jorgensd opened the pull request:
This PR is a starting point for modernizing the ufl code base.

We start by introducing:
- [Type hints PEP 484](https://www.python.org/dev/peps/pep-0484/)
- [fstrings PEP 498](https://www.python.org/dev/peps/pep-0498/) 


Also several functions that has been marked as deprecated has been removed:
- `Argument.ufl_domain()`
- `Argument.ufl_domains()`
- `Argument.ufl_element()`

jorgensd commented on ufl/argument.py:
```diff
@@ -38,97 +40,74 @@ class Argument(FormArgument):
-            domain = default_domain(element.cell())
-            function_space = FunctionSpace(domain, element)
-        elif not isinstance(function_space, AbstractFunctionSpace):
+        if not isinstance(function_space, AbstractFunctionSpace):
```
```suggestion
        if isinstance(function_space, FiniteElementBase):
            # For legacy support for .ufl files using cells, we map the cell to
            # the default Mesh
            element = function_space
            domain = default_domain(element.cell())
            function_space = FunctionSpace(domain, element)
        elif not isinstance(function_space, AbstractFunctionSpace):
```

wence- commented on ufl/algorithms/analysis.py:
```diff
@@ -143,8 +143,10 @@ def extract_arguments_and_coefficients(a):
+    arguments, coefficients = extract_arguments_and_coefficients(form)
+    elements_a = tuple(a.ufl_function_space().ufl_element() for a in arguments)
+    elements_c = tuple(c.ufl_element() for c in coefficients)
+    return elements_a + elements_c
```
Why don’t coefficients have a function space?

jorgensd replied:
It does. A mistake on my behalf.

wence- commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -545,13 +545,21 @@ def argument(self, o):
+            el = f.ufl_function_space().ufl_element()
+        else:
+            el = f.ufl_element()
+        if el.mapping() == "physical":
```
Why does this backwards compat (?) need to be introduced?

wence- commented on ufl/argument.py:
```diff
@@ -54,81 +59,63 @@ def __init__(self, function_space, number, part=None):
-    def ufl_element(self):
-        "Deprecated, please use .ufl_function_space().ufl_element() instead."
-        # TODO: deprecate("Argument.ufl_domain() is deprecated, please
-        # use .ufl_function_space().ufl_element() instead.")
```
TBH, I think this deprecation comment seems misguided. I think that there should be a consistent way to get the domain from any expression, and it should probably be `ufl_domains`. Otherwise you have to special case arguments everywhere.

wence- replied:
what was the resolution to this?

wence- commented on ufl/domain.py:
```diff
@@ -339,7 +339,11 @@ def extract_domains(expr):
+        if hasattr(t, "ufl_function_space"):
+            domains = t.ufl_function_space().ufl_domains()
+        else:
+            domains = t.ufl_domains()
```
See, this seems like a backwards step.

wence- reviewed (changes requested):
I think most of the cleanup looks fine, but the removal of ufl_domain from arguments looks like a step in the wrong direction.

jorgensd commented:
> I think most of the cleanup looks fine, but the removal of ufl_domain from arguments looks like a step in the wrong direction.

I've reverted the removal of `ufl_domains` (but still removed `ufl_element` and `ufl_domain`. 
I've also cleaned up the code using the fact that coefficients also have a `ufl_function_space`.

jhale commented on ufl/algebra.py:
```diff
@@ -318,7 +299,7 @@ def evaluate(self, x, mapping, component, index_values):
 
     def __str__(self):
         a, b = self.ufl_operands
-        return "%s ** %s" % (parstr(a, self), parstr(b, self))
```
Result looks different to old code?

jorgensd replied:
It should give the same result. MWE:
```python
import ufl

cell = ufl.triangle
el = ufl.VectorElement("Lagrange", cell, 1)
mesh = ufl.Mesh(el)
a = ufl.Constant(mesh)
b = ufl.Constant(mesh)
c = a ** b
print(c)
```
yielding 
```bash
c_0 ** c_1
```

jhale replied:
You're right! I haven't used join a great deal. Is it really clearer than an f-string or format? I can see the advantage for many arguments, but 2 seems marginal.

jorgensd replied:
> You're right! I haven't used join a great deal. Is it really clearer than an f-string or format? I can see the advantage for many arguments, but 2 seems marginal.

I agree it is marginal. I’ll just revert it.

wence- commented on ufl/algorithms/apply_derivatives.py:
```diff
@@ -545,13 +545,16 @@ def argument(self, o):
-        domain = f.ufl_domain()
+
+        domains = f.ufl_function_space().ufl_domains()
+        domain = None if len(domains) == 0 else domains[0]
```
I think it should be the case that a function space always has a domain? Should we instead write this as:

```
domain, = f.ufl_function_space().ufl_domains()
```

To assert that there is only one.

wence- commented on ufl/algorithms/analysis.py:
```diff
@@ -143,8 +143,9 @@ def extract_arguments_and_coefficients(a):
 
 def extract_elements(form):
     "Build sorted tuple of all elements used in form."
+    arguments, coefficients = extract_arguments_and_coefficients(form)
```
this seems unneeded now?

wence- commented on ufl/algorithms/compute_form_data.py:
```diff
@@ -60,8 +60,9 @@ def _compute_element_mapping(form):
-    elements = [obj.ufl_element() for obj in chain(form.arguments(),
-                                                   form.coefficients())]
+    elements_a = tuple(a.ufl_function_space().ufl_element() for a in form.arguments())
+    elements_c = tuple(c.ufl_element() for c in form.coefficients())
```
coefficients and arguments both have function spaces attached, so can both these iterations be merged?

wence- reviewed (commented):
I'm a bit confused by some of the changes in this PR. Most of it is just some minor formatting changes for consistency, but then there are some actual API changes that I think warrant discussion in terms of whether they are a step forward (eg the stuff around arguments and domains). Can you split it in two to make it easier to review?

mscroggs commented:
We should maybe close this, as I've started doing this is a series of smaller PRs (starting with #131)
